File: packages/@vizij/arora-types/src/messages.ts
/**
 * Arora WebSocket Message Types
 *
 * This module defines the message format for WebSocket communication with the
 * Arora API, matching the `arora-websocket` crate (1.0). Messages are
 * serialized as JSON with a `type` field discriminator, and speak the Arora
 * data-layer vocabulary: values are written to and read from **keys**
 * (hierarchical paths into the store).
 */

import type { AroraType, AroraValue } from "./value";

// ============================================================================
// Metadata Types
// ============================================================================

/**
 * Metadata describing a key exposed by the runtime's data layer.
 * Keys represent controllable parameters or observable outputs.
 */
export type KeyInfo = {
  /** Hierarchical path identifier (e.g., "face/mouth/open") */
  path: string;
  /** Key kind/category (e.g., "input", "output", "computed") */
  kind?: string;
  /** The arora Type of the values this key accepts/produces */
  value_type?: AroraType;
  /** Minimum value constraint (for numeric types) */
  min?: number;
  /** Maximum value constraint (for numeric types) */
  max?: number;
  /** Default value */
  default_value?: AroraValue;
  /** Human-readable description */
  description?: string;
};

/**
 * Descriptor for an RPC method parameter.
 */
export type MethodParam = {
  /** Parameter name */
  name: string;
  /** Parameter type */
  param_type: AroraType;
  /** Whether this parameter is required */
  required?: boolean;
  /** Default value if not provided */
  default_value?: AroraValue;
  /** Human-readable description */
  description?: string;
};

/**
 * Metadata describing an available RPC method.
 */
export type MethodInfo = {
  /** Method path/name (e.g., "audio/play", "animation/trigger", "reset") */
  path: string;
  /** Method parameters */
  params?: MethodParam[];
  /** Return type (undefined means void/unit) */
  return_type?: AroraType;
  /** Human-readable description */
  description?: string;
};

// ============================================================================
// Incoming Messages (Client -> Server)
// ============================================================================

/**
 * Messages sent from client to server.
 * Use the `type` field to discriminate message kind.
 */
export type Incoming =
  | { type: "write_values"; values: Record<string, AroraValue> }
  | { type: "read_values"; keys: string[] }
  | { type: "list_keys"; path?: string }
  | { type: "list_methods"; path?: string }
  | {
      type: "invoke";
      method: string;
      args?: Record<string, AroraValue>;
      request_id?: string;
    };

// ============================================================================
// Outgoing Messages (Server -> Client)
// ============================================================================

/**
 * Messages sent from server to client.
 * Use the `type` field to discriminate message kind.
 */
export type Outgoing =
  | { type: "write_values_resp"; success: boolean; message?: string }
  | { type: "read_values_resp"; values: Record<string, AroraValue> }
  | { type: "list_keys_resp"; keys: KeyInfo[] }
  | { type: "list_methods_resp"; methods: MethodInfo[] }
  | {
      type: "invoke_resp";
      success: boolean;
      request_id?: string;
      value?: AroraValue;
      message?: string;
    }
  | { type: "error"; request_id?: string; message: string }
  | { type: "values_changed"; values: Record<string, AroraValue> };

// ============================================================================
// Type Guards
// ============================================================================

/** Check if message is a WriteValuesResp */
export function isWriteValuesResp(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "write_values_resp" }> {
  return msg.type === "write_values_resp";
}

/** Check if message is a ReadValuesResp */
export function isReadValuesResp(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "read_values_resp" }> {
  return msg.type === "read_values_resp";
}

/** Check if message is a ListKeysResp */
export function isListKeysResp(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "list_keys_resp" }> {
  return msg.type === "list_keys_resp";
}

/** Check if message is a ListMethodsResp */
export function isListMethodsResp(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "list_methods_resp" }> {
  return msg.type === "list_methods_resp";
}

/** Check if message is an InvokeResp */
export function isInvokeResp(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "invoke_resp" }> {
  return msg.type === "invoke_resp";
}

/** Check if message is an Error */
export function isError(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "error" }> {
  return msg.type === "error";
}

/** Check if message is a ValuesChanged push */
export function isValuesChanged(
  msg: Outgoing,
): msg is Extract<Outgoing, { type: "values_changed" }> {
  return msg.type === "values_changed";
}

// ============================================================================
// Message Constructors
// ============================================================================

/**
 * Create a WriteValues message.
 */
export function createWriteValues(
  values: Record<string, AroraValue>,
): Incoming {
  return { type: "write_values", values };
}

/**
 * Create a ReadValues message.
 */
export function createReadValues(keys: string[]): Incoming {
  return { type: "read_values", keys };
}

/**
 * Create a ListKeys message.
 */
export function createListKeys(path?: string): Incoming {
  return { type: "list_keys", path };
}

/**
 * Create a ListMethods message.
 */
export function createListMethods(path?: string): Incoming {
  return { type: "list_methods", path };
}

/**
 * Create an Invoke message.
 */
export function createInvoke(
  method: string,
  args?: Record<string, AroraValue>,
  request_id?: string,
): Incoming {
  return { type: "invoke", method, args, request_id };
}

File: packages/arora-websocket/src/messages.rs
//! WebSocket message types bridging the Arora API.
//!
//! Messages are serialized as JSON with a `type` field discriminator, and
//! speak the Arora data-layer vocabulary: values are written to and read from
//! **keys** (hierarchical paths into the store). This is the arora-websocket
//! 1.0 wire format.

use arora_connection::{KeyInfo, MethodInfo, Value};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Messages received from WebSocket clients.
///
/// All incoming messages use a `type` field to discriminate the message kind.
/// Example JSON: `{"type": "write_values", "values": {...}}`
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Incoming {
    /// Write values to keys.
    ///
    /// Example: `{"type": "write_values", "values": {"face/mouth": {"f64": 0.5}}}`
    WriteValues {
        /// Map of key paths to their new values
        values: HashMap<String, Value>,
    },

    /// Read the current values of keys.
    ///
    /// Example: `{"type": "read_values", "keys": ["face/mouth", "face/eyes"]}`
    ReadValues {
        /// List of key paths to retrieve
        keys: Vec<String>,
    },

    /// Request the list of available keys.
    ///
    /// Example: `{"type": "list_keys"}` or `{"type": "list_keys", "path": "face"}`
    ListKeys {
        /// Optional path prefix to filter keys
        #[serde(default)]
        path: Option<String>,
    },

    /// Request the list of available RPC methods.
    ///
    /// Example: `{"type": "list_methods"}` or `{"type": "list_methods", "path": "audio"}`
    ListMethods {
        /// Optional path prefix to filter methods
        #[serde(default)]
        path: Option<String>,
    },

    /// Invoke an RPC method.
    ///
    /// Example: `{"type": "invoke", "method": "reset", "request_id": "req-1"}`
    Invoke {
        /// Method path/name to invoke
        method: String,

        /// Arguments as key-value pairs
        #[serde(default)]
        args: HashMap<String, Value>,

        /// Optional request ID for correlating responses
        #[serde(default)]
        request_id: Option<String>,
    },
}

/// Messages sent to WebSocket clients.
///
/// All outgoing messages use a `type` field to discriminate the message kind.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Outgoing {
    /// Response to WriteValues message.
    ///
    /// Example: `{"type": "write_values_resp", "success": true}`
    WriteValuesResp {
        /// Whether the write was successful
        success: bool,

        /// Error message if success is false
        #[serde(skip_serializing_if = "Option::is_none")]
        message: Option<String>,
    },

    /// Response to ReadValues message.
    ///
    /// Example: `{"type": "read_values_resp", "values": {"face/mouth": {"f64": 0.5}}}`
    ReadValuesResp {
        /// Map of key paths to their current values
        values: HashMap<String, Value>,
    },

    /// Response to ListKeys message.
    ///
    /// Example: `{"type": "list_keys_resp", "keys": [...]}`
    ListKeysResp {
        /// List of matching keys
        keys: Vec<KeyInfo>,
    },

    /// Response to ListMethods message.
    ///
    /// Example: `{"type": "list_methods_resp", "methods": [...]}`
    ListMethodsResp {
        /// List of matching methods
        methods: Vec<MethodInfo>,
    },

    /// Response to Invoke message.
    ///
    /// Example: `{"type": "invoke_resp", "success": true, "request_id": "req-1"}`
    InvokeResp {
        /// Whether the invocation was successful
        success: bool,

        /// Echo back request_id if provided in the request
        #[serde(skip_serializing_if = "Option::is_none")]
        request_id: Option<String>,

        /// Return value from the method (if any)
        #[serde(skip_serializing_if = "Option::is_none")]
        value: Option<Value>,

        /// Error message if success is false
        #[serde(skip_serializing_if = "Option::is_none")]
        message: Option<String>,
    },

    /// Generic error response for parse failures or unknown message types.
    ///
    /// Example: `{"type": "error", "message": "Invalid JSON"}`
    Error {
        /// Echo back request_id if available
        #[serde(skip_serializing_if = "Option::is_none")]
        request_id: Option<String>,

        /// Error description
        message: String,
    },

    /// Server-initiated push: values changed (e.g. the runtime wrote new
    /// state). Sent unsolicited to subscribed clients.
    ///
    /// Example: `{"type": "values_changed", "values": {"face/mouth": {"f64": 0.5}}}`
    ValuesChanged {
        /// Map of key paths to their new values.
        values: HashMap<String, Value>,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_incoming_write_values_deserialize() {
        let json = r#"{"type": "write_values", "values": {"test/path": {"f64": 0.5}}}"#;
        let msg: Incoming = serde_json::from_str(json).unwrap();
        match msg {
            Incoming::WriteValues { values } => {
                assert!(values.contains_key("test/path"));
            }
            _ => panic!("Expected WriteValues message"),
        }
    }

    #[test]
    fn test_incoming_read_values_deserialize() {
        let json = r#"{"type": "read_values", "keys": ["face/mouth", "face/eyes"]}"#;
        let msg: Incoming = serde_json::from_str(json).unwrap();
        match msg {
            Incoming::ReadValues { keys } => {
                assert_eq!(keys, vec!["face/mouth", "face/eyes"]);
            }
            _ => panic!("Expected ReadValues message"),
        }
    }

    #[test]
    fn test_incoming_list_keys_deserialize() {
        let json = r#"{"type": "list_keys", "path": "face"}"#;
        let msg: Incoming = serde_json::from_str(json).unwrap();
        match msg {
            Incoming::ListKeys { path } => {
                assert_eq!(path, Some("face".to_string()));
            }
            _ => panic!("Expected ListKeys message"),
        }
    }

    #[test]
    fn test_incoming_invoke_deserialize() {
        let json = r#"{"type": "invoke", "method": "reset", "request_id": "req-1"}"#;
        let msg: Incoming = serde_json::from_str(json).unwrap();
        match msg {
            Incoming::Invoke {
                method,
                args,
                request_id,
            } => {
                assert_eq!(method, "reset");
                assert!(args.is_empty());
                assert_eq!(request_id, Some("req-1".to_string()));
            }
            _ => panic!("Expected Invoke message"),
        }
    }

    #[test]
    fn test_outgoing_write_values_resp_serialize() {
        let msg = Outgoing::WriteValuesResp {
            success: true,
            message: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"write_values_resp""#));
        assert!(json.contains(r#""success":true"#));
        assert!(!json.contains("message"));
    }

    #[test]
    fn test_outgoing_invoke_resp_serialize() {
        let msg = Outgoing::InvokeResp {
            success: false,
            request_id: Some("req-1".to_string()),
            value: None,
            message: Some("Method not found".to_string()),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"invoke_resp""#));
        assert!(json.contains(r#""request_id":"req-1""#));
        assert!(json.contains(r#""message":"Method not found""#));
    }

    #[test]
    fn test_outgoing_values_changed_serialize() {
        let values: HashMap<String, Value> =
            serde_json::from_str(r#"{"face/mouth": {"f64": 0.5}}"#).unwrap();
        let msg = Outgoing::ValuesChanged { values };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains(r#""type":"values_changed""#));
        assert!(json.contains(r#""face/mouth""#));
    }
}
